src/safemode_ui.cpp

```diff
@@ -0,0 +1,740 @@
+                                                      " " + get_options().get_option( "SAFEMODEPROXIMITY" ).getDefaultText(),
+                                                      "",
+                                                      3,
+                                                      true
+                                                    );
+                if( text.empty() ) {
+                    current_tab[line].proximity = get_option<int>( "SAFEMODEPROXIMITY" );
+                } else {
+                    //Let the options class handle the validity of the new value
+                    auto temp_option = get_options().get_option( "SAFEMODEPROXIMITY" );
+                    temp_option.setValue( text.c_str() );
+                    current_tab[line].proximity = atoi( temp_option.getValue().c_str() );
+                }
+            }
+        } else if( action == "ENABLE_RULE" && current_tab_non_empty ) {
+            changes_made = true;
+            current_tab[line].active = true;
+        } else if( action == "DISABLE_RULE" && current_tab_non_empty ) {
+            changes_made = true;
+            current_tab[line].active = false;
+        } else if( action == "LEFT" ) {
+            column--;
+            if( column < 0 ) {
+                column = num_columns - 1;
+            }
+        } else if( action == "RIGHT" ) {
+            column++;
+            if( column >= num_columns ) {
+                column = 0;
+            }
+        } else if( action == "MOVE_RULE_UP" && current_tab_non_empty ) {
+            changes_made = true;
+            if( line < ( int )current_tab.size() - 1 ) {
+                std::swap( current_tab[line], current_tab[line + 1] );
+                line++;
+                column = 0;
+            }
+        } else if( action == "MOVE_RULE_DOWN" && current_tab_non_empty ) {
+            changes_made = true;
+            if( line > 0 ) {
+                std::swap( current_tab[line],  current_tab[line - 1] );
+                line--;
+                column = 0;
+            }
+        } else if( action == "TEST_RULE" && current_tab_non_empty ) {
+            test_pattern( tab, line );
+        } else if( action == "SWITCH_SAFEMODE_OPTION" ) {
+            get_options().get_option( "SAFEMODE" ).setNext();
+            get_options().save();
+        }
+    }
+
+    if( !changes_made ) {
+        return;
+    }
+
+    if( query_yn( _( "Save changes?" ) ) ) {
+        if( is_safemode_in ) {
+            save_global();
+            if( g->u.name != "" ) {
+                save_character();
+            }
+        } else {
+            create_rules();
+        }
+    } else {
+        global_rules = global_rules_old;
+        character_rules = character_rules_old;
+    }
+}
+
+void safemode::test_pattern( const int tab_in, const int row_in )
+{
+    std::vector<std::string> creature_list;
+    std::string creature_name = "";
```

Comment:
I'm wondering why people write this explicit initialization for strings, but use the default constructor for the vector (above). The ` = ""` is entirely redundant here.